Custom visibility timeout for each message in SQS queue

tejasraorane opened the pull request:
First the visibility timeout for each message was set to 0. Now we can set a custom value from 0 to 43200 seconds.12 hours is the maximum. We just need to set it in spring configuration as a property of AmazonAwsSQSConnectorFactory. name="visibilityTimeoutOnReset" value="10"

Commit:
Update AbstractSQSConnector.java

To add custom visibility timeout for each message in SQS queue

Commit:
Update AmazonAwsSQSConnectorFactory.java

Add custom visibility timeout to each message in SQS queue.You can set the value of visibilityTimeoutOnReset in spring configuration. Value is in seconds.

Commit:
Update AmazonAwsSQSConnector.java

To set custom visibility timeout for each message in SQS queue.

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/AbstractSQSConnector.java:
```diff
@@ -34,16 +34,18 @@
+        this(receiveCheckIntervalMs, false, 0);
     }
 
-    protected AbstractSQSConnector(long receiveCheckIntervalMs, boolean isAsync)
```
Since this method isn't private, you can't just change it's signature.  You'll have to add a third constructor.

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/AbstractSQSConnector.java:
```diff
@@ -34,16 +34,18 @@
 
     private final long _receiveCheckIntervalMs;
     private final boolean _isAsync;
+    private int _visibilityTimeoutOnReset;
```
Make final.  Also add a getter.

carterpage replied:
Still needs to be final.

carterpage replied:
You'll need to make this final.

tejasraorane replied:
When i make it final, the constructor does allow me to do that. Remove final modifier error is shown.

carterpage replied:
That's because you need to fix the constructor.

On Fri, Jul 4, 2014 at 3:36 AM, tejas1990 notifications@github.com wrote:

> In
> nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/AbstractSQSConnector.java:
> 
> > @@ -34,16 +34,18 @@
> > 
> > ```
> >  private final long _receiveCheckIntervalMs;
> >  private final boolean _isAsync;
> > ```
> > -    private int _visibilityTimeoutOnReset;
> 
> When i make it final, the constructor does allow me to do that. Remove
> final modifier error is shown.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/skyscreamer/nevado/pull/86/files#r14549770.

tejasraorane replied:
Can you please suggest how can i do that. Cause your constructor is giving error that the _visibilityTimeoutOnReset may not have been initialized.

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/amazonaws/AmazonAwsSQSConnector.java:
```diff
@@ -50,12 +50,12 @@
 
-    public AmazonAwsSQSConnector(String awsAccessKey, String awsSecretKey, boolean isSecure, long receiveCheckIntervalMs) {
-        this(awsAccessKey, awsSecretKey, isSecure, receiveCheckIntervalMs, false);
+    public AmazonAwsSQSConnector(String awsAccessKey, String awsSecretKey, boolean isSecure, long receiveCheckIntervalMs, , int visibilityTimeoutOnReset) {
```
You'll have to add new constructors.  You can't change signatures on non-private methods.

Also, this line seems to have an extra comma.  Did you compile and also run the tests?

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/amazonaws/AmazonAwsSQSConnectorFactory.java:
```diff
@@ -10,11 +10,12 @@
  */
 public class AmazonAwsSQSConnectorFactory extends AbstractSQSConnectorFactory {
     protected boolean _useAsyncSend = false;
+    protected int visibilityTimeoutOnReset;
```
Please prefix with "_"

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/amazonaws/AmazonAwsSQSConnectorFactory.java:
```diff
@@ -31,4 +32,12 @@ public void setUseAsyncSend(boolean useAsyncSend) {
+	}
+
+	public void setVisibilityTimeoutOnReset(int visibilityTimeoutOnReset) {
+		this.visibilityTimeoutOnReset = visibilityTimeoutOnReset;
```
No "this" needed.  Just use _visibilityTimeoutOnReset

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/amazonaws/AmazonAwsSQSConnectorFactory.java:
```diff
@@ -31,4 +32,12 @@ public void setUseAsyncSend(boolean useAsyncSend) {
+    
+    public int getVisibilityTimeoutOnReset() {
+		return visibilityTimeoutOnReset;
+	}
```
Fix indentation to match rest of file.

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/amazonaws/AmazonAwsSQSConnectorFactory.java:
```diff
@@ -31,4 +32,12 @@ public void setUseAsyncSend(boolean useAsyncSend) {
+
+	public void setVisibilityTimeoutOnReset(int visibilityTimeoutOnReset) {
+		this.visibilityTimeoutOnReset = visibilityTimeoutOnReset;
+	}
```
Fix indentation.

tejasraorane committed:
Update AmazonAwsSQSConnectorFactory.java

Made changes as per your suggestions.

tejasraorane committed:
Update AmazonAwsSQSConnector.java

Made changes as per your suggestions. Created a new constructor.

tejasraorane committed:
Update AbstractSQSConnector.java

Made changes as per your suggestions. I was not able to set visibilitytimeout as final as it was causing error in your current method- 
protected AbstractSQSConnector(long receiveCheckIntervalMs, boolean isAsync)
    {
        _receiveCheckIntervalMs = receiveCheckIntervalMs;
        _isAsync = isAsync;
    }

tejasraorane committed:
Update AmazonAwsSQSConnector.java

removed unwanted constructor.

tejasraorane commented:
I have made changes as per your suggestions.please review it.

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/AbstractSQSConnector.java:
```diff
@@ -45,10 +47,20 @@ protected AbstractSQSConnector(long receiveCheckIntervalMs, boolean isAsync)
     }
+    
+    public int getVisibilityTimeoutOnReset() {
+		return _visibilityTimeoutOnReset;
```
Fix indentation.

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/AbstractSQSConnector.java:
```diff
@@ -127,7 +139,9 @@ public void resetMessage(NevadoMessage message) throws JMSException {
         }
         SQSQueue sqsQueue = getSQSQueue(message.getNevadoDestination());
-        sqsQueue.setMessageVisibilityTimeout(sqsReceiptHandle, 0);
+        if (sqsReceiptHandle != null && !BLANK_STRING.equals(sqsReceiptHandle)){
```
Unnecessary.  sqsReceiptHandle can't be null.  Don't do the BLANK_STRING check either.

tejasraorane replied:
But above this line you have done a null check for sqsReceiptHandle.

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/AbstractSQSConnector.java:
```diff
@@ -29,11 +29,13 @@
  */
 public abstract class AbstractSQSConnector implements SQSConnector {
     protected static final String AWS_ERROR_CODE_AUTHENTICATION = "InvalidClientTokenId";
+    protected static final String BLANK_STRING = "";
```
Per below we don't need this.  Regardless, this is a poor way to check for an empty string.  Using Apache commons StringUtils.isEmpty() or (s == null || s.trim().length() == 0) is better.

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/AbstractSQSConnector.java:
```diff
@@ -179,7 +193,9 @@ protected SQSMessage receiveSQSMessage(NevadoConnection connection, NevadoDestin
                     try {
-                        sqsQueue.setMessageVisibilityTimeout(sqsMessage.getReceiptHandle(), 0); // Make it immediately available to the next requestor
+                        if (sqsMessage.getReceiptHandle() != null && !BLANK_STRING.equals(sqsMessage.getReceiptHandle())) {
+                    		sqsQueue.setMessageVisibilityTimeout(sqsMessage.getReceiptHandle(), _visibilityTimeoutOnReset); // Customize message visibility timeout
```
Per above, don't do this.  (The "if" check.)

tejasraorane replied:
Sometimes it throws an exception "Unable to reset visibility timeout as the receipthandle does not exist". That is why i made this check.

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/amazonaws/AmazonAwsSQSConnector.java:
```diff
@@ -51,11 +51,11 @@
 
-    public AmazonAwsSQSConnector(String awsAccessKey, String awsSecretKey, boolean isSecure, long receiveCheckIntervalMs, boolean isAsync) {
-        super(receiveCheckIntervalMs, isAsync);
+    public AmazonAwsSQSConnector(String awsAccessKey, String awsSecretKey, boolean isSecure, long receiveCheckIntervalMs, boolean isAsync, int visibilityTimeoutOnReset) {
```
You cannot change the signature on a public method.

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/amazonaws/AmazonAwsSQSConnectorFactory.java:
```diff
@@ -31,4 +32,12 @@ public void setUseAsyncSend(boolean useAsyncSend) {
     }
+    
+   public int get_visibilityTimeoutOnReset() {
+		return _visibilityTimeoutOnReset;
```
Fix indentation.

Also getVisibilityTimeoutOnReset.  No "_".

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/amazonaws/AmazonAwsSQSConnectorFactory.java:
```diff
@@ -31,4 +32,12 @@ public void setUseAsyncSend(boolean useAsyncSend) {
+		return _visibilityTimeoutOnReset;
+	}
+
+	public void set_visibilityTimeoutOnReset(int visibilityTimeoutOnReset) {
```
Same here.

carterpage commented:
Needs more work.

tejasraorane committed:
Update AbstractSQSConnector.java

Removed BLANK_STRING as you suggested. I have kept the if conditions as sometimes it throws an exception "Unable to reset visibility timeout as receipthandle does not exist in queue" even if the message is there in the queue.

tejasraorane committed:
Update AbstractSQSConnector.java

Correction made to if conditions for visibility timeout.

tejasraorane committed:
Update AmazonAwsSQSConnector.java

Made changes as per your suggestions. Created my own constructor.

tejasraorane committed:
Update AmazonAwsSQSConnectorFactory.java

Made changes as per your suggestions.

tejasraorane commented:
Made changes as per your suggestions. I have kept the if checks as sometimes it throws exception "Unable to reset message visibility as the receipt handle does not exist" even if the message is present in the queue.

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/AbstractSQSConnector.java:
```diff
@@ -127,7 +138,9 @@ public void resetMessage(NevadoMessage message) throws JMSException {
         }
         SQSQueue sqsQueue = getSQSQueue(message.getNevadoDestination());
-        sqsQueue.setMessageVisibilityTimeout(sqsReceiptHandle, 0);
+        if (sqsReceiptHandle != null && StringUtils.isNotEmpty(sqsReceiptHandle) && sqsReceiptHandle.trim().length() != 0){
```
You're scaring me now.

tejasraorane replied:
Should I remove if checks or not. Because when i do load testing on my code it gives the exception for many messages. For single message it is fine but for many messages sometimes it throws exception.

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/amazonaws/AmazonAwsSQSConnectorFactory.java:
```diff
@@ -10,11 +10,12 @@
  */
 public class AmazonAwsSQSConnectorFactory extends AbstractSQSConnectorFactory {
     protected boolean _useAsyncSend = false;
+    protected int _visibilityTimeoutOnReset;
```
Need a default.

carterpage commented on nevado-jms/src/main/java/org/skyscreamer/nevado/jms/connector/amazonaws/AmazonAwsSQSConnector.java:
```diff
@@ -76,6 +76,29 @@ public AmazonAwsSQSConnector(String awsAccessKey, String awsSecretKey, boolean i
             _amazonSNS = new AmazonSNSClient(awsCredentials, clientConfiguration);
         }
     }
+    
```
Cut and paste anti-pattern.

http://en.wikipedia.org/wiki/Copy_and_paste_programming

carterpage commented:
This PR is too far from completion, and I'm no longer comfortable with adding this to the code base.  This library is used by thousands, and no one else has expressed the visibility setting as being a problem.  Nevado expects to operate on dedicated queues that are not being used for other purposes.   It manually changes the visibility timeouts on messages to be compliant with the JMS spec, and changing that is not within the nature of the project.
